adoptingerlang.org

```diff
@@ -4026,18 +4101,118 @@ No business works only with code; it is way more frequent to hear about projects
+
+- build engineering becomes more complex as more tools and build chains need support
+- more dependencies need auditing and upkeep to keep everything current
+- as time goes on, some languages become uncool and require more effort to train, hire for, or rewrite
+- we're doing polyglot and there's this one component nobody wants to touch because a dev once wrote it in a language nobody else knows
+
+Build engineering is a legit concern, in the same way many organisations with a dedicated IT department ships out a single line of computers for every developer to reduce the surface area of what is required. That being said, this is a less prevalent issue with the advent of pervasive containerization. Docker images and runtimes such as Kubernetes already provide a one-size-fits-all interface that extends from local development to production runs. Maintaining build images with the help of people familiar with the programming languages on board mostly reduce the complexity of everything involved over time.
+
+There isn't necessarily a great way to work around dependency explosion and monitoring, short of using tools that scan and report issues for you. That being said, dependency surface area is mostly a problem of managing complexity, and isn't necessarily tied to how many languages you used. For example, we've seen single javascript demo projects with dependency trees larger than established Erlang software used for entire departments in a company.
+
+If your company has a procedure where each dependency must be vetted for a security audit, legal department (for licensing purposes), and engineers (for code quality), you'll find out that in practice, it doesn't work great. People will either stick to a pre-whitelisted list of libraries, will only review top-level dependencies. In some cases, they'll never update outdated libraries because they prefer to avoid all the red tape. In some cases devs will just rewrite things, to avoid having to cross many departments of reviews if they feel they gotta ship things fast, and you'll end up with code that barely gets reviewed at all. We consider this to be an organisational issue, not a programming language issue.
+
+Uncool languages or languages that are "hard to hire for" can always be a problem. Unfortunately, this is a problem you can _never_ actually solve, because nobody knows the future. Is it worse to have a single main language people stop liking and then none of your stack is easy to staff, or to have only half of it being hard to hire for? What about cases like Python, where the migration from 2.x to 3.x has been so hard that libraries and components are essentially considered to be distinct languages? We consider that organisations that do not marry themselves to a single language but instead make a habit of switching between a few of them (and not just a ton of them) train themselves to better adjust to variations in the market of programming languages.
```

Comment:
I tend not to like "best tool for the job" because it's a bit of a catch-all for "I like this thing", probably something to cover in the defending erlang chapter. Kind of want a space to expand on this rant here: https://twitter.com/mononcqc/status/1199303744791416834